packages/animations/lib/src/modal.dart

```diff
@@ -0,0 +1,208 @@
+  /// Creates a [_ModalRoute] route with the Material fade transition.
+  ///
+  /// [barrierDismissible] is true by default.
+  _ModalRoute({
+    Color barrierColor,
+    bool barrierDismissible = true,
+    String barrierLabel,
+    _ModalTransitionBuilder transitionBuilder,
+    Duration transitionDuration,
+    Duration reverseTransitionDuration,
+    @required this.builder,
+  })  : assert(barrierDismissible != null),
+        _barrierColor = barrierColor,
+        _barrierDismissible = barrierDismissible,
+        _barrierLabel = barrierLabel,
+        _transitionBuilder = transitionBuilder,
+        _transitionDuration = transitionDuration,
+        _reverseTransitionDuration = reverseTransitionDuration;
+
+  @override
+  Color get barrierColor => _barrierColor;
+  final Color _barrierColor;
+
+  @override
+  bool get barrierDismissible => _barrierDismissible;
+  final bool _barrierDismissible;
+
+  @override
+  String get barrierLabel => _barrierLabel;
+  final String _barrierLabel;
+
+  @override
+  Duration get transitionDuration => _transitionDuration;
+  final Duration _transitionDuration;
+
+  @override
+  Duration get reverseTransitionDuration => _reverseTransitionDuration;
+  final Duration _reverseTransitionDuration;
+
+  /// The primary contents of the modal.
+  final WidgetBuilder builder;
+
+  @override
+  Widget buildPage(
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+  ) {
+    final ThemeData theme = Theme.of(context);
+    return Semantics(
+      child: SafeArea(
+        child: Builder(
+          builder: (BuildContext context) {
+            final Widget child = Builder(builder: builder);
+            return theme != null ? Theme(data: theme, child: child) : child;
+          },
+        ),
+      ),
+      scopesRoute: true,
+      explicitChildNodes: true,
+    );
+  }
+
+  @override
+  Widget buildTransitions(
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    return _transitionBuilder(
+      context,
+      animation,
+      secondaryAnimation,
+      child,
+    );
+  }
+
+  final _ModalTransitionBuilder _transitionBuilder;
+}
+
+/// A configuration object containing the properties needed to implement a
+/// modal route.
+///
+/// The `barrierDismissible` argument is used to determine whether this route
+/// can be dismissed by tapping the modal barrier. This argument defaults
+/// to true. If `barrierDismissible` is true, a non-null `barrierLabel` must be
+/// provided.
+///
+/// The `barrierLabel` argument is the semantic label used for a dismissible
+/// barrier. This argument defaults to "Dismiss".
+abstract class ModalConfiguration {
+  /// Creates a modal configuration object that provides the necessary
+  /// properties to implement a modal route.
+  ModalConfiguration({
+    this.barrierColor,
+    this.barrierDismissible,
```

Comment:
Can `barrierDismissible`, `transitionDuration` and `reverseTransitionDuration` legally be null? If not, assert and document?

Reply:
barrierDismissible legally cannot be null. Since `transitionDuration` and `reverseTransitionDuration` are typically set by overrides to the properties, those can remain as-is